File: chrome-extension/lib/summarize.ts
import Anthropic from "@anthropic-ai/sdk";
import schema from "../../shared/schema.json";
import promptText from "../../shared/prompt.md?raw";
import type { TranscriptSegment } from "./transcript";
import type { Chapter } from "./chapters";

/** One section of the generated summary. */
export interface Section {
  /** Section start, in seconds — matches a real transcript-segment start time. */
  sec: number;
  title: string;
  summary: string;
}

/** The full structured output Claude produces. */
export interface Summary {
  /** ISO 639-1 code of the language Claude wrote in. */
  language: string;
  sections: Section[];
}

/** Minimal shape of the Anthropic client we depend on (for testability). */
export interface AnthropicLike {
  messages: {
    create(body: Record<string, unknown>): Promise<{ content: unknown[] }>;
  };
}

export interface SummarizeInput {
  /** Normalized transcript segments — see `normalizeJson3`. */
  transcript: TranscriptSegment[];
  /** Creator-defined chapter markers, if any — see `extractChapters`. */
  chapters?: Chapter[] | null;
  model: string;
  apiKey: string;
}

export interface SummarizeDeps {
  createAnthropic?: (apiKey: string) => AnthropicLike;
  /** Injected in tests to avoid real timers. Defaults to setTimeout-based sleep. */
  sleep?: (ms: number) => Promise<void>;
}

const MAX_TOKENS = 4096;

/**
 * Orchestrate one summary: build the prompt from the transcript, force Claude
 * to call the `submit_summary` tool, and return the validated summary. How the
 * transcript bytes are obtained is the caller's concern.
 */
export async function summarize(
  input: SummarizeInput,
  deps: SummarizeDeps = {},
): Promise<Summary> {
  const createAnthropic = deps.createAnthropic ?? defaultCreateAnthropic;

  const client = createAnthropic(input.apiKey);
  const message = await client.messages.create({
    model: input.model,
    max_tokens: MAX_TOKENS,
    system: promptText,
    tools: [schema],
    tool_choice: { type: "tool", name: "submit_summary" },
    messages: [{ role: "user", content: renderUserMessage(input) }],
  });

  const summary = parseSummary(message.content);
  return snapTimestamps(summary, input.transcript.map((s) => s.sec));
}

function renderUserMessage(input: SummarizeInput): string {
  let message = "";

  if (input.chapters && input.chapters.length > 0) {
    const list = input.chapters.map((ch) => `[${ch.sec}] ${ch.title}`).join("\n");
    message += `Video chapters:\n\n${list}\n\n`;
  }

  const body = input.transcript.map((s) => `[${s.sec}] ${s.text}`).join("\n");
  message += `Summarize this video transcript:\n\n${body}`;
  return message;
}

function parseSummary(content: unknown[]): Summary {
  const toolUse = content.find(
    (block): block is { type: "tool_use"; name: string; input: unknown } =>
      isToolUse(block) && block.name === "submit_summary",
  );
  if (!toolUse) {
    throw new Error("Claude did not call submit_summary");
  }
  return validateSummary(toolUse.input);
}

function isToolUse(
  block: unknown,
): block is { type: string; name: string; input: unknown } {
  return (
    typeof block === "object" &&
    block !== null &&
    (block as { type?: unknown }).type === "tool_use"
  );
}

function validateSummary(input: unknown): Summary {
  const value = input as Partial<Summary>;
  if (typeof value?.language !== "string") {
    throw new Error("submit_summary: missing language");
  }
  if (!Array.isArray(value.sections) || value.sections.length === 0) {
    throw new Error("submit_summary: sections must be a non-empty array");
  }
  for (const section of value.sections) {
    if (
      typeof section?.sec !== "number" ||
      typeof section?.title !== "string" ||
      typeof section?.summary !== "string"
    ) {
      throw new Error("submit_summary: malformed section");
    }
  }
  return value as Summary;
}

export function snapTimestamps(summary: Summary, realSegmentStarts: number[]): Summary {
  const sorted = [...realSegmentStarts].sort((a, b) => a - b);
  return {
    ...summary,
    sections: summary.sections.map((section) => {
      // No segment ≤ sec (fabricated timestamp before the transcript starts): clamp to first.
      const snapped = findFloor(sorted, section.sec) ?? sorted[0];
      return snapped === section.sec ? section : { ...section, sec: snapped };
    }),
  };
}

function findFloor(sorted: number[], target: number): number | undefined {
  let result: number | undefined;
  for (const val of sorted) {
    if (val <= target) result = val;
    else break;
  }
  return result;
}

/**
 * Like `summarize`, but retries exactly once after a 2-second delay when
 * Anthropic returns a 429 (rate-limit). Any other error is rethrown immediately.
 */
export async function summarizeWithRetry(
  input: SummarizeInput,
  deps: SummarizeDeps = {},
): Promise<Summary> {
  const sleep = deps.sleep ?? defaultSleep;
  try {
    return await summarize(input, deps);
  } catch (err) {
    if (isHttpStatus(err, 429)) {
      await sleep(2000);
      return summarize(input, deps);
    }
    throw err;
  }
}

function isHttpStatus(err: unknown, status: number): boolean {
  return typeof (err as any)?.status === "number" && (err as any).status === status;
}

function defaultSleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

function defaultCreateAnthropic(apiKey: string): AnthropicLike {
  return new Anthropic({
    apiKey,
    dangerouslyAllowBrowser: true,
  }) as unknown as AnthropicLike;
}

File: ios-app/YTSummarizer/Sources/YTSummarizerCore/Summarize.swift
import Foundation

// MARK: - Constants (API key replaced in NIM-14)

let claudeAPIKey = ProcessInfo.processInfo.environment["ANTHROPIC_API_KEY"] ?? ""
let claudeModel  = "claude-sonnet-4-6"

// MARK: - Timestamp snapping

/// Snap every section.sec to the nearest real transcript segment start ≤ sec.
/// Sections before the first segment are clamped to the first segment.
///
/// Mirrors `snapTimestamps` in `chrome-extension/lib/summarize.ts`.
func snapTimestamps(summary: Summary, realSegmentStarts: [Int]) -> Summary {
    let sorted = realSegmentStarts.sorted()
    guard !sorted.isEmpty else { return summary }
    let snappedSections = summary.sections.map { section -> Section in
        let snapped = findFloor(sorted: sorted, target: section.sec) ?? sorted[0]
        guard snapped != section.sec else { return section }
        return Section(sec: snapped, title: section.title, summary: section.summary)
    }
    return Summary(language: summary.language, sections: snappedSections)
}

private func findFloor(sorted: [Int], target: Int) -> Int? {
    var result: Int?
    for val in sorted {
        if val <= target { result = val } else { break }
    }
    return result
}

// MARK: - Response parsing

/// Extract and validate a `Summary` from a raw Anthropic Messages API response dictionary.
func parseSummaryResponse(_ response: [String: Any]) throws -> Summary {
    guard let content = response["content"] as? [[String: Any]],
          let toolUse = content.first(where: {
              $0["type"] as? String == "tool_use" &&
              $0["name"] as? String == "submit_summary"
          }),
          let input = toolUse["input"] as? [String: Any] else {
        throw SummarizeError.missingToolUse
    }

    guard let language = input["language"] as? String else {
        throw SummarizeError.missingLanguage
    }

    guard let sectionsRaw = input["sections"] as? [[String: Any]], !sectionsRaw.isEmpty else {
        throw SummarizeError.missingSections
    }

    let sections = try sectionsRaw.map { s -> Section in
        guard let sec   = s["sec"]     as? Int,
              let title = s["title"]   as? String,
              let sum   = s["summary"] as? String else {
            throw SummarizeError.malformedSection
        }
        return Section(sec: sec, title: title, summary: sum)
    }

    return Summary(language: language, sections: sections)
}

// MARK: - Claude API call

/// Call the Anthropic Messages API, force a `submit_summary` tool call,
/// and return the snapped summary.
func callClaude(transcript: [TranscriptSegment]) async throws -> Summary {
    guard let promptURL = Bundle.module.url(forResource: "prompt", withExtension: "md"),
          let schemaURL = Bundle.module.url(forResource: "schema",  withExtension: "json") else {
        throw SummarizeError.apiError("prompt.md or schema.json not found in bundle")
    }

    let promptText = try String(contentsOf: promptURL, encoding: .utf8)
    let schemaData = try Data(contentsOf: schemaURL)
    guard let schema = try? JSONSerialization.jsonObject(with: schemaData) as? [String: Any] else {
        throw SummarizeError.apiError("schema.json is not a valid JSON object")
    }

    let userMessage = renderUserMessage(transcript: transcript)
    let body: [String: Any] = [
        "model":      claudeModel,
        "max_tokens": 4096,
        "system":     promptText,
        "tools":      [schema],
        "tool_choice": ["type": "tool", "name": "submit_summary"],
        "messages":   [["role": "user", "content": userMessage]]
    ]
    let bodyData = try JSONSerialization.data(withJSONObject: body)

    var request = URLRequest(url: URL(string: "https://api.anthropic.com/v1/messages")!)
    request.httpMethod = "POST"
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    request.setValue(claudeAPIKey,       forHTTPHeaderField: "x-api-key")
    request.setValue("2023-06-01",       forHTTPHeaderField: "anthropic-version")
    request.httpBody = bodyData

    let (data, response) = try await URLSession.shared.data(for: request)
    guard let http = response as? HTTPURLResponse else {
        throw SummarizeError.apiError("Unexpected response type")
    }
    guard http.statusCode == 200 else {
        let body = String(data: data, encoding: .utf8) ?? "(empty body)"
        throw SummarizeError.apiError("HTTP \(http.statusCode): \(body)")
    }

    guard let responseJSON = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
        throw SummarizeError.apiError("Response is not a JSON object")
    }

    let summary = try parseSummaryResponse(responseJSON)
    return snapTimestamps(summary: summary, realSegmentStarts: transcript.map { $0.sec })
}

// MARK: - Helpers

private func renderUserMessage(transcript: [TranscriptSegment]) -> String {
    let body = transcript.map { "[\($0.sec)] \($0.text)" }.joined(separator: "\n")
    return "Summarize this video transcript:\n\n\(body)"
}
